src/osvr/USBSerial/USBSerialDevInfo_MacOSX.h

```diff
@@ -87,15 +85,62 @@ namespace usbserial {
 
             return (vendor_matches && product_matches);
         }
-
-    } // end namespace
+    }
 
     std::vector<USBSerialDevice>
     getSerialDeviceList(boost::optional<uint16_t> vendorID,
                         boost::optional<uint16_t> productID) {
 
-        //FIXME: STUB
         std::vector<USBSerialDevice> devices;
+
+        io_iterator_t serialPortIterator;
+        io_object_t serialPortService;
+        // find all serial ports
+        if (findSerialPorts(&serialPortIterator) == true) {
+            // iterate over serial ports, getting vid and pid
+            while ((serialPortService = IOIteratorNext(serialPortIterator)) !=
+                   0) {
+                const CFNumberRef vidObj =
+                    static_cast<CFNumberRef>(IORegistryEntrySearchCFProperty(
+                        serialPortService, kIOServicePlane, CFSTR("idVendor"),
+                        NULL, kIORegistryIterateRecursively |
+                                  kIORegistryIterateParents));
+                const CFNumberRef pidObj =
+                    static_cast<CFNumberRef>(IORegistryEntrySearchCFProperty(
+                        serialPortService, kIOServicePlane, CFSTR("idProduct"),
+                        NULL, kIORegistryIterateRecursively |
+                                  kIORegistryIterateParents));
+                const CFStringRef bsdPathObj =
+                    static_cast<CFStringRef>(IORegistryEntryCreateCFProperty(
+                        serialPortService, CFSTR(kIOCalloutDeviceKey),
+                        kCFAllocatorDefault, 0));
+
+                if (!vidObj || !pidObj || !bsdPathObj) {
+                    continue;
+                }
+                // handle device
+                uint16_t vid, pid;
+                CFNumberGetValue(vidObj, kCFNumberSInt16Type, &vid);
+                CFNumberGetValue(pidObj, kCFNumberSInt16Type, &pid);
+
+                // convert the string object into a C-string
+                CFIndex bufferSize = CFStringGetMaximumSizeForEncoding(
+                                         CFStringGetLength(bsdPathObj),
+                                         kCFStringEncodingMacRoman) +
+                                     sizeof('\0');
+                std::vector<char> bsdPathBuf(bufferSize);
+                CFStringGetCString(bsdPathObj, &bsdPathBuf[0], bufferSize,
+                                   kCFStringEncodingMacRoman);
+                // create the device
```

Comment:
That's annoying documentation - they heavily imply that the output gets null-terminated, but don't explicitly say so. Ah well, par for the course for string functions - for instance <https://randomascii.wordpress.com/2013/04/03/stop-using-strncpy-already/>